src/sage/schemes/hyperelliptic_curves_smooth_model/hyperelliptic_g2.py

```diff
@@ -0,0 +1,293 @@
+            return (c0**2 + 4 * a0) == 0
+
+    @cached_method
+    def jacobian(self):
+        r"""
+        Return the Jacobian of the hyperelliptic curve.
+
+        Elements of the Jacobian are represented by tuples
+        of the form `(u, v : n)`, where
+
+        - `(u,v)` is the Mumford representative of a divisor `P_1 + ... + P_r`,
+        - `n` is a non-negative integer
+
+        This tuple represents the equivalence class
+        `[P_1 + ... + P_r + n \cdot \infty_+ + m \cdot \infty_- - D_\infty]`,
+        where  `m = g - \deg(u) - n`, and `\infty_+`, `\infty_-` are the
+        points at infinity of the hyperelliptic curve,
+
+        .. MATH::
+
+            D_\infty =
+            \lceil g/2 \rceil \infty_+ + \lfloor g/2 \rfloor \infty_-.
+
+        Here, `\infty_- = \infty_+`, if the hyperelliptic curve is ramified.
+
+
+        EXAMPLES::
+
+            sage: R.<x> = QQ[]
+            sage: H = HyperellipticCurveSmoothModel(2*x^5 + 4*x^4 + x^3 - x, x^3 + x + 1)
+            sage: J = H.jacobian(); J
+            Jacobian of Hyperelliptic Curve over Rational Field defined by y^2 + (x^3 + x + 1)*y = 2*x^5 + 4*x^4 + x^3 - x
+
+        The points `P = (0, 0)` and `Q = (-1, -1)` are on `H`. We construct the
+        element `D_1 = [P - Q] = [P + (-Q) - D_\infty]` on the Jacobian::
+
+            sage: P = H.point([0, 0])
+            sage: Q = H.point([-1, -1])
+            sage: D1 = J(P,Q); D1
+            (x^2 + x, -2*x : 0)
+
+        Elements of the Jacobian can also be constructed by directly providing
+        the Mumford representation::
+
+            sage: D1 == J(x^2 + x, -2*x, 0)
+            True
+
+        We can also embed single points into the Jacobian. Below we construct
+        `D_2 = [P - P_0]`, where `P_0` is the distinguished point of `H`
+        (by default one of the points at infinity)::
+
+            sage: D2 = J(P); D2
+            (x, 0 : 0)
+            sage: P0 = H.distinguished_point(); P0
+            (1 : 0 : 0)
+            sage: D2 == J(P, P0)
+            True
+
+        We may add elements, or multiply by integers::
+
+            sage: 2*D1
+            (x, -1 : 1)
+            sage: D1 + D2
+            (x^2 + x, -1 : 0)
+            sage: -D2
+            (x, -1 : 1)
+
+        Note that the neutral element is given by `[D_\infty - D_\infty]`,
+        in particular `n = 1`::
+
+            sage: J.zero()
+            (1, 0 : 1)
+
+        There are two more elements of the Jacobian that are only supported
+        at infinity: `[\infty_+ - \infty_-]` and `[\infty_- - \infty_+]`::
+
+            sage: [P_plus, P_minus] = H.points_at_infinity()
+            sage: P_plus == P0
+            True
+            sage: J(P_plus,P_minus)
+            (1, 0 : 2)
+            sage: J(P_minus, P_plus)
+            (1, 0 : 0)
+        """
+
+        from sage.schemes.hyperelliptic_curves_smooth_model.jacobian_g2_generic import (
+            HyperellipticJacobian_g2_generic,
+        )
```

Comment:
It's better to have all imports at the top of the file at the module level unless they are expensive or cause circular import issues.